### PR TITLE
i2c: nrfx-twim: enable pm runtime from the code

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Change all nrf twi instances to enable PM runtime using pm_device_runtime_enable() rather than the zephyr,pm-device-runtime-auto property.

The nrf TWI devices can use three different drivers: twi, twim and twis, only twim uses the PM runtime APIs internally.

Right now the nRF52 soc dtsi files set twim and enables PM runtime, but then the nRF5 DK files override the driver to twi but do not remove the zephyr,pm-device-runtime-auto property. This combination results in the twi driver initializing in suspended state and never being resumed.

Enabling PM runtime from the driver code itself resolves the situation and removes the ambiguity.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
@JordanYates I think I remember you mentioning that a device should not call `pm_device_runtime_enable` on itself, but in this case it seems to make sense as doing it from dt with the device supporting three different drivers seems like a potential source of problems, what do you think?

**Comment by JordanYates** [comment]
> @JordanYates I think I remember you mentioning that a device should not call `pm_device_runtime_enable` on itself, but in this case it seems to make sense as doing it from dt with the device supporting three different drivers seems like a potential source of problems, what do you think?

I did say that.
The reasoning is that the user should be able to control at init time whether PM is enabled or not.
That is currently achieved by the existence of `zephyr,pm-device-runtime-auto` on the devicetree node.
If the driver calls `pm_device_runtime_enable` on itself, the user has no option to stop this happening except to manually call `pm_device_runtime_disable` in the application, with all the problems this would cause for board agnostic applications.

The other solution that has been floated, all devices default to enabling PM if `CONFIG_PM_DEVICE_RUNTIME=y` with a DT property to disable the behaviour, preserves the ability of the user to change the init time behaviour, but is obviously a larger change. 2 weeks before feature freeze means it is probably too late to do for 4.0, but doing it at the start of the 4.1 cycle may be a better option? As far as I can see this PR is a minor cleanup of the Nordic `.dtsi` files and so wouldn't hurt from being bumped to 4.1.

**Review by gmarull — CHANGES_REQUESTED** [review]
we should likely fix twi driver instead, let me check.

**Comment by fabiobaltieri** [comment]
> If the driver calls `pm_device_runtime_enable` on itself, the user has no option to stop this happening except to manually call `pm_device_runtime_disable` in the application, with all the problems this would cause for board agnostic applications.

Sure but this driver also uses the PM apis on itself, wouldn't it make sense to keep the enable as well in the code? The issue here is that if you build for any nRF52 devkit with PM runtime the i2c is just not going to work.

> we should likely fix twi driver instead, let me check.

Yeah the alternative is to use

```
compatible = "nordic,nrf-twi";
/delete-property/ zephyr,pm-device-runtime-auto;
```

but that feels backward

**Comment by JordanYates** [comment]
> Sure but this driver also uses the PM apis on itself, wouldn't it make sense to keep the enable as well in the code? The issue here is that if you build for any nRF52 devkit with PM runtime the i2c is just not going to work.

Whether the driver uses the other PM APIs on itself doesn't matter, because the outcome of those calls is the same regardless of whether `CONFIG_PM_DEVICE_RUNTIME` is enabled or not, and whether PM is enabled on that particular instance or not. Specifically, a call to `i2c_transfer` "just works".

**Comment by fabiobaltieri** [comment]
> Whether the driver uses the other PM APIs on itself doesn't matter, because the outcome of those calls is the same regardless of whether `CONFIG_PM_DEVICE_RUNTIME` is enabled or not, and whether PM is enabled on that particular instance or not. Specifically, a call to `i2c_transfer` "just works".

Yeah but `i2c_nrfx_twi.c` does not use ANY PM API on itself, at all. So if you enable PM runtime on it `i2c_transfer` does not work anymore, and it's very easy to get in that situation because of how the devicetree files are organized now.

**Comment by fabiobaltieri** [comment]
@JordanYates let me rephrase the issue in a different way, I think I'm not explaining myself very well:

```
$ west build -t flash -p -b nrf52dk/nrf52832 samples/subsys/shell/shell_module/ -DCONFIG_I2C=y -DCONFIG_I2C_SHELL=y
...
uart:~$ i2c scan i2c@40003000 
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:             -- -- -- -- -- -- -- -- -- -- -- -- 
 <works>
```

```
$ west build -t flash -p -b nrf52dk/nrf52832 samples/subsys/shell/shell_module/ -DCONFIG_I2C=y -DCONFIG_I2C_SHELL=y -DCONFIG_PM_DEVICE=y -DCONFIG_PM_DEVICE_RUNTIME=y
...
uart:~$ i2c scan i2c@40003000                                                   
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f                             
00:             -- -- -- -- --
<hangs, every i2c transaction times out>
uart:~$ device list                                                             
devices:                                                                        
- clock@40000000 (READY)                                                        
- gpio@50000000 (READY)                                                         
- uart@40002000 (active)                                                        
- i2c@40003000 (suspended, usage=0)                                             
```


```
$ git diff
diff --git a/boards/nordic/nrf52dk/nrf52dk_nrf52832.dts b/boards/nordic/nrf52dk/nrf52dk_nrf52832.dts
index 458f5918b32..8a4fe9cea7a 100644
--- a/boards/nordic/nrf52dk/nrf52dk_nrf52832.dts
+++ b/boards/nordic/nrf52dk/nrf52dk_nrf52832.dts
@@ -169,6 +169,7 @@ arduino_i2c: &i2c0 {
        pinctrl-0 = <&i2c0_default>;
        pinctrl-1 = <&i2c0_sleep>;
        pinctrl-names = "default", "sleep";
+       /delete-property/ zephyr,pm-device-runtime-auto;
 };
 
 &i2c1 {
$ west build -t flash -p -b nrf52dk/nrf52832 samples/subsys/shell/shell_module/ -DCONFIG_I2C=y -DCONFIG_I2C_SHELL=y -DCONFIG_PM_DEVICE=y -DCONFIG_PM_DEVICE_RUNTIME=y
...
$uart:~$ i2c scan i2c@40003000                                                   
<works again>
```

How should this be handled? Either we tell the user to delete the property or we move the enable to the driver, we certainly can't tell the user to do a runtime get of the i2c device, that makes no sense.

**Comment by fabiobaltieri** [comment]
hey @gmarull you want me to close this? really I don't care too much just think the current setup is a bit of a trap

**Comment by kartben** [comment]
upcoming PR dashboard says this needs a rebase 😗

**Comment by fabiobaltieri** [comment]
> upcoming PR dashboard says this needs a rebase 😗

ah it'll probably get closed, waiting to hear if Gerard has any better ideas
